lib/cracker.rb

```diff
@@ -1,3 +1,61 @@
-class Cracker
+require './lib/decoder'
 
+class Cracker < Decoder
+
+  def reverse_shifts(ciphertext, known_end = " end")
+    last_four = ciphertext[-known_end.length..-1].split('')
+    last_four.zip(known_end.split('')).map do |cipher_char, known_char|
+      shift = @whitelist.index(cipher_char) - @whitelist.index(known_char)
+      shift.negative? ? (shift + @whitelist.length) : shift
+    end.rotate(-(ciphertext.length % known_end.length))
+  end
+
+  def root_codes(shifts, offsets)
+    shifts.zip(offsets).map {|shift, offset| shift - offset.to_i}
+  end
+
+  def full_shift(code)
+    code + @whitelist.length
+  end
+
+  def code_to_string(code)
+    code.to_s.length < 2 ? code.to_s.prepend("0") : code.to_s
+  end
+
+  def potential_codes(root_codes)
+    root_codes.map do |code|
+      codes = [code]
+      codes << full_shift(codes.last) until full_shift(codes.last) > 99
+      codes.map {|potential_code| code_to_string(potential_code)}
+    end
+  end
+
+  def all_combinations(code_arrays)
+    code_arrays[0].product(code_arrays[1], code_arrays[2], code_arrays[3])
```

Comment:
product is an array method that returns an array of all combinations of elements from all arrays and it's basically the reason iteration 4 happened